### PR TITLE
added Polynomial cutoff function as descibed by Peterson and Khorshidi

### PR DESCRIPTION
**Pull request by EricMusa**
*Includes derivatives. 
Tested with basic script --> compared gamma=2. polynomial function against cosine function (similar forms, as depicted in fig. 5 of ([Peterson & Khorshidi](https://www.sciencedirect.com/science/article/pii/S0010465516301266?via%3Dihub)) --> similar values and derivative values, everything checks out

Next step is to look into adding control flow in the symmetry function code to use either the cosine cutoff function or the polynomial function for calculations

### CODE REVIEW

**Comment by mshuaibii** [comment]
Thanks for the PR @EricMusa. Lets merge this once we get the control flow in place and try training some models with it being used. If you'd like to work build off this that would be great.

**Review by EricMusa — COMMENTED** [review]
**Modifications for choosing Polynomial vs Cosine cutoff function made:**
 - I've added the "cutoff_params" dict to the "descriptor_setup" tuple of descriptor settings.
 - "cutoff_params" by default will be {'cutoff_func': 'cosine'} but can also be of the form {'cutoff_func': 'polynomial', 'gamma': <float>}. 
 - This get's passed to the Gaussian descriptor object which, when using _libsymf for calculations, now calls either "calculate_sf_cos" or "calculate_sf_poly" (or their noderiv complements) depending on the passed cutoff function. 
 - I've modified "example.py" to create two models - each with a descriptor set using one of the two cutoff functions - and then print the accuracies of the two models. 
 - The script runs fine and the results are very similar.

**Review by EricMusa — COMMENTED** [review]
couple of bug fixed and more testing:
 - modified Gaussian.py get_descriptor_setup_hash to include info about the cutoff function
    - no longer overwriting same Gs with diff cutoff functions

**Comment by mshuaibii** [comment]
Awesome - can you please merge master into this. I've included some CI tests so lets make sure this looks fine with that. If you haven't already done so installing the `pre-commit` hooks (`black` specifically) will take care of formatting/structure to be consistent with the rest of the repo. That's usually the thing people fail on in the tests, just an FYI.

**Comment by mshuaibii** [comment]
> couple of bug fixed and more testing:
> 
> * modified Gaussian.py get_descriptor_setup_hash to include info about the cutoff function
>   
>   * no longer overwriting same Gs with diff cutoff functions

Awesome - can you please merge master into this. I've included some CI tests so lets make sure this looks fine with that. If you haven't already done so installing the `pre-commit` hooks (`black` specifically) will take care of formatting/structure to be consistent with the rest of the repo. That's usually the thing people fail on in the tests, just an FYI.

**Review by EricMusa — COMMENTED** [review]
`black` pre-commit hook and trim-tailing-whitespace reformatted a lot of the files, also made some various edits to imports to pass `flake8` hook as well.

**Comment by mshuaibii** [comment]
Can you run something like this - I had CI builds off for forked-PRs, should be fixed now: `git commit --allow-empty -m "trigger build"`

File: amptorch/tests/consistency_test.py [hunk 1]
```diff
@@ -121,7 +120,7 @@ def test_energy_force_consistency():
     torch_Cu_weights_2 = torch.FloatTensor(Cu_weights_2[:-1, :]).t()
     torch_Cu_bias_2 = torch.FloatTensor(Cu_weights_2[-1, :])
 
-    descriptor_setup = ("gaussian", Gs, elements)
+    descriptor_setup = ("gaussian", Gs, {"cutoff_func": "cosine"}, elements)
```
Comment (mshuaibii) [inline amptorch/tests/consistency_test.py:123]: This descriptor setup is starting to look a little awkward with the nested dictionary. Can we do this differently/cleaner so it's more clear what things are?

```
descriptor_setup  = {
    "descriptor": Gaussian,
    "params": Gs,
    "cutoff_func": ...
    "elements":...
}
```

**Review by mshuaibii — CHANGES_REQUESTED** [review]

**Comment by mshuaibii** [comment]
Looks good - minor changes to make things more clear/easier to understand.

**Comment by mshuaibii** [comment]
@EricMusa FYI you only needed to `trigger build` once, you should be fine to push your commits and builds fine.

**Comment by EricMusa** [comment]
Yea, sorry about that, I'm learning as I go haha
Changes:
- reverted example.py back to the upstream/master branch's version (as it was before)
- fixed nested dict in consistency_test.py
- added cutoff_funcs_test.py to test Cosine CF vs Polynomial CF (gamma=2.0)

**Review by mshuaibii — APPROVED** [review]

**Pull request merged**